generic3g/Restart.F90

```diff
@@ -0,0 +1,195 @@
+   use mapl3g_pFIOServerBounds, only: pFIOServerBounds
+   use mapl3g_SharedIO, only: esmf_to_pfio_type
+   use MAPL_FieldPointerUtilities, only: FieldGetCPtr, FieldGetLocalElementCount
+   use pFIO, only: PFIO_READ, FileMetaData, NetCDF4_FileFormatter
+   use pFIO, only: i_Clients, o_Clients, ArrayReference
+
+   implicit none
+   private
+
+   public :: Restart
+
+   type :: Restart
+      private
+      character(len=ESMF_MAXSTR) :: gc_name
+      type(ESMF_Geom) :: gc_geom
+      type(ESMF_Time) :: current_time
+   contains
+      procedure, public :: write
+      procedure, public :: read
+      procedure, private :: write_bundle_
+      procedure, private :: read_fields_
+   end type Restart
+
+   interface Restart
+      procedure new_Restart
+   end interface Restart
+
+contains
+
+   function new_Restart(gc_name, gc_geom, gc_clock, rc) result(new_rstrt)
+      character(len=*), intent(in) :: gc_name
+      type(ESMF_Geom), intent(in) :: gc_geom
+      type(ESMF_Clock), intent(in) :: gc_clock
+      integer, optional, intent(out) :: rc
+      type(Restart) :: new_rstrt ! result
+
+      integer :: status
+
+      new_rstrt%gc_name = ESMF_UtilStringLowerCase(trim(gc_name), _RC)
+      call ESMF_Clockget(gc_clock, currTime = new_rstrt%current_time, _RC)
+      new_rstrt%gc_geom = gc_geom
+
+      _RETURN(ESMF_SUCCESS)
+   end function new_Restart
+
+   subroutine write(this, state_type, state, rc)
+      ! Arguments
+      class(Restart), intent(inout) :: this
+      character(len=*), intent(in) :: state_type
+      type(ESMF_State), intent(in) :: state
+      integer, optional, intent(out) :: rc
+
+      ! Locals
+      type(ESMF_FieldBundle) :: out_bundle
+      character(len=ESMF_MAXSTR) :: file_name
+      integer :: item_count, status
+
+      call ESMF_StateGet(state, itemCount=item_count, _RC)
+      if (item_count > 0) then
+         file_name = trim(this%gc_name) // "_" // trim(state_type) // "_checkpoint.nc4"
+         print *, "Writing checkpoint: ", trim(file_name)
+         out_bundle = get_bundle_from_state_(state, _RC)
+         call this%write_bundle_(out_bundle, file_name, rc)
+      end if
+
+      _RETURN(ESMF_SUCCESS)
+   end subroutine write
+
+   subroutine read(this, state_type, state, rc)
+
+      ! Arguments
+      class(Restart), intent(inout) :: this
+      character(len=*), intent(in) :: state_type
+      type(ESMF_State), intent(inout) :: state
+      integer, optional, intent(out) :: rc
+
+      ! Locals
+      character(len=ESMF_MAXSTR) :: file_name
+      logical :: file_exists
+      integer :: item_count, status
+
+      call ESMF_StateGet(state, itemCount=item_count, _RC)
+      if (item_count > 0) then
+         file_name = trim(this%gc_name) // "_" // trim(state_type) // "_rst.nc4"
+         inquire(file=trim(file_name), exist=file_exists)
+         if (file_exists) then
+            print *, "Reading restart: ", trim(file_name)
+            call this%read_fields_(file_name, state, _RC)
+         else
```

Comment:
Bonus points for eliminating else.

Reply:
I understand.   I'm tending to agree with a lot of developers that early return is preferable to `else` in terms of code clarity.  When this is not the case, it usually indicates other improvements that can be made to the procudere. (New helper procedure for example.)